### PR TITLE
Use PartialObject for object predicates in _.every, _.find and _.some 

### PR DESCRIPTION
**Pull request by lukas-zech-software**
This PR introduces the new Typescript keyword *keyof* to make the predicate objects used in _.every, _.find and _.some partially typed to their respective collections' type

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

`tslint` returns a few hundred warnings, which were already present before this PR
To keep this PR scoped to the actual changes I did not fix any lint warnings other than those affected by my changes

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: no changes made to API
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

The changes only make use of a the new Typescript  keyword *keyof* the meaning of the typings haven't changed at all

### CODE REVIEW

**Comment by lukas-zech-software** [comment]
This is the follow up merge request for https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13985 which I closed to get the branches right

**Comment by sheetalkamat** [comment]
@lukas-zech-software  or @andy-ms Can you please take a look at failing tests.
Thank you

**Comment by lukas-zech-software** [comment]
> Package compiles in TypeScript 2.1 but not in 2.0.
You can add a line '// TypeScript Version: 2.1' to the end of the header to specify a new compiler version.


The tests seems to fail for Typescript 2.0 which is to be expected, as the PR uses the **keyof** operator which is only available in >= TS 2.1 

Thats why I added the line **// TypeScript Version: 2.1** in the header of the file, but the tests do not seem to respect that and test it against TS 2.0 anyway.

Is there a way to stop the test from trying to compile agains TS 2.0 ?

**Comment by sheetalkamat** [comment]
@andy-ms will know more about test framework.

File: lodash/index.d.ts [hunk 1]
```diff
@@ -242,6 +243,10 @@ export as namespace _;
 declare var _: _.LoDashStatic;
 
 declare namespace _ {
+    type PartialObject<T> = {
```
Comment (mhegazy) [inline lodash/index.d.ts]: this is just `Partial<T>`, it is already defined in lib.d.ts, why redefine it?

**Review by mhegazy — CHANGES_REQUESTED** [review]

**Comment by mhegazy** [comment]
Can you merge https://github.com/DefinitelyTyped/DefinitelyTyped/pull/14066 into this PR?

**Comment by lukas-zech-software** [comment]
@mhegazy 
merged #14066 into this

I added `type PartialObject` as there is already an interface called `Partial` in this file for lodash's `partial` function.

The name conflicted and I could not find any way to address the `Partial` from `lib.es6.d.ts`.

`global.Partial` does not work and I did not want to rename the already implemented `Partial` interface as this would break the typings API.

Do you know of a better way to import types from `lib.es6.d.ts` explicitly?

**Comment by mhegazy** [comment]
> Do you know of a better way to import types from lib.es6.d.ts explicitly?

outside the namespace `_`, declare a type `type PartialObject<T> = Partial<T>;` then use that instead.

**Comment by mhegazy** [comment]
looks like tests are still failing.

**Comment by lukas-zech-software** [comment]
@andy-ms could you help me out with the tests?
They still fail and I can't see why.

Error is
`../lodash/index.d.ts(245,25): error TS2304: Cannot find name 'Partial'.`

which would indicate that the tests do not run with TS 2.1 as the `Partial` type is defined in `lib.es6.d.ts` of TS 2.1 and should be available globally.


At the end it still says 
```
Package compiles in TypeScript 2.1 but not in 2.0.
You can add a line '// TypeScript Version: 2.1' to the end of the header to specify a new compiler version.
```

which is the expected outcome. why is this still considered failed?

**Comment by andy-ms** [comment]
This is because `lodash` isn't the package being tested, another package is. So that will have to be updated to use TS2.1 too.
The affected packages are `bookshelf`, `lodash-decorators`, `lodash-es`, `sequelize/v3`, `sequelize`, `sequelize-fixtures`, and `umzug`.

**Comment by lukas-zech-software** [comment]
I added the header to all failing packages and most work now, but `lodash-decorators` still fails.

When I run `tsc -p ./` locally in the `lodash-decorators` everything compiles without an error.

I get the same errors as in the travis log when compiling the test file without the correct tsconfig.json like so: `tsc  lodash-decorators-tests.ts`

So it seems like the test runner does not use the tsconfig from the `lodash-decorators` folder?

I can't see how to resolve this, any suggestions?

File: lodash/index.d.ts (hunk 1 above)
Comment (lukas-zech-software) [inline lodash/index.d.ts]: I redefined type outside the namespace like you suggested

**Comment by andy-ms** [comment]
I think you just need to merge from `master`. Fixed the tsconfig recently (#14111).

**Comment by lukas-zech-software** [comment]
Finally all tests are green 😃 

@mhegazy can you review the changes you request?

**Comment by lukas-zech-software** [comment]
@mhegazy any updates on this?

**Comment by ericanderson** [comment]
:+1:

**Comment by jkmart** [comment]
Now getting `/build/node_modules/@types/lodash/index.d.ts(245,25): error TS2304: Cannot find name 'Partial'.` when building against `typescript: "~2.0.3"`

**Comment by andy-ms** [comment]
You can install `@types/lodash@ts2.0` if you want to keep using TypeScript 2.0.

**Comment by pgonzal** [comment]
Could we revert this PR, and redo it as a minor or major SemVer increment (or at least delay it until TS2.1 has wider adoption)?

For many developers, upgrading to the TypeScript 2.1 compiler is a nontrivial amount of work.  My team spent nearly two weeks getting our SPFx framework upgraded to TypeScript 2.1, and now we have to go through various rings of validation before we can push out this change to our consumers, some of whom may have to go through a similar migration process themselves.

A major lesson was that a compiler migration must happen in a specific order:  first you do the applications, then the middleware, then the core libraries, and finally the typings.  Otherwise, any package using 2.1 syntax becomes unusable by projects that are still on 2.0.  Thus, a ubiquitous library like **lodash** should be the *last* package to upgrade, not an early adopter.

In any case, it seems inappropriate to introduce such a disruptive change as a minor SemVer increment (4.14.50 -> 4.14.51).  We could try to workaround this issue by locking the version number in our **package.json** (i.e. specify "4.14.50" instead of "~4.14.38"), but this dependency is shared across 20+ different packages in different Git repos, so it would inevitably lead to side-by-side installs of different versions of the typings.  We can't upgrade everything all at once.

**Comment by ericanderson** [comment]
@pgonzal I understand your frustration, I've recently heard from some people who felt similarly to the changes we made to React along these same lines. We have quite the long discussion here https://github.com/DefinitelyTyped/DefinitelyTyped/issues/14250 and of relevance may be my final thoughts https://github.com/DefinitelyTyped/DefinitelyTyped/issues/14250#issuecomment-275399347

Ultimately, most people benefit from these changes. TS 2.0 -> 2.1 is a straight forward upgrade. The code should mostly compile and the few places it doesn't have in my experience uncovered bugs that 2.0 happily allowed through and 2.1 prevented.

It has been almost 2 months since 2.1 was first released and the folks at Microsoft and DefinitelyTyped put a strict 30 day block on introducing changes that would require 2.1. Further, the entire publishing infrastructure has changed to support old versions to be tagged as stated above by @andy-ms. 

In order for TypeScript to continue to grow and become the best tool for developing web-based software (something I once fought strongly against within my organization), it has to be able to move fast. Specifically some of the new types in 2.1 were created specifically to better support things like lodash and react.

No one is forcing you to use DefinitelyTyped. I (published an article)[https://ericlanderson.com/using-custom-typescript-definitions-with-ts-2-x-3121db84015d#.3jofjw83v] on Dec 7 to help people to use the latest and greatest features until DefinitelyTyped was updated. That same article can be used to rewrite these definitions for old versions too if needed.

In fact, your organization could publish your types for lodash once and depend strictly on those if you must control them and keep them in sync.

Ultimately, I am not a decider here though and happily defer to those who do choose the path for DefinitelyTyped, but I definitely believe pushing forward is whats best for the community.

**Comment by pgonzal** [comment]
> Ultimately, most people benefit from these changes. TS 2.0 -> 2.1 is a straight forward 
> upgrade. The code should mostly compile and the few places it doesn't have in my experience
> uncovered bugs that 2.0 happily allowed through and 2.1 prevented.

As I mentioned above, it took several people *two weeks* of continuous work to migrate us from TypeScript 2.0 to 2.1.  Our projects are split across four Git repositories, comprising 76 different NPM packages (including NodeJS toolchain packages, browser libraries, and several different end-user facing applications).  One of our deliverables is the SPFx framework, which third party developers obtain by running "npm install".  Our Yeoman developer experience is currently in a [broken state](https://github.com/SharePoint/sp-dev-docs/issues/394#issuecomment-275953859) due to this PR, and won't have a proper fix until we push out a new release of the framework, which is typically an entire day of publishing, validation, writing up announcements, etc.

Here are some of the problems we encountered with TypeScript 2.1:

- Code that compiled under 2.0 fails to compile with 2.1 (e.g. type checks behave differently)

- Code that executed correctly under 2.0 can mysteriously malfunction at runtime when built with 2.1 (e.g. [if you inherit from the Error class](https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes))

- The compiler API changed, meaning tools like **tslint** don’t work correctly unless replaced with a new, compatible version, which requires config files to be updated.  We maintain such a tool ("**@microsoft/api-extractor**"), and someone had to invest a couple days of coding to get it working with the TypeScript 2.1 compiler API.  We also have custom **ts-lint** rules.

- If a library exposes a TypeScript 2.1 keyword, this forces all the consumers to upgrade to the new version.  We cannot simply upgrade all 75 projects in a single commit (e.g. the build tools are used to build themselves), so we have to carefully choreograph the order in which this happens.  The third parties, whose code we cannot see, are part of this dependency graph.

- For internal packages, we simply did a major SemVer increment when they migrated to TypeScript 2.1, which avoided build breaks.  But for third party packages, we cannot do this, because a major SemVer increment would cause side-by-side loading for shipped components, effectively doubling the amount of script code being loaded in the browser.  Instead, we need to think of the "developer typings" version as being a separate kind of breaking change, that is distinct from a runtime SemVer change.

When we upgraded from TypeScript 1.0 -> 2.0, it all went pretty smoothly, since it was a carefully planned effort across multiple partner teams.  But TypeScript 2.0 -> 2.1 was really painful, since people assumed it was simple "minor" update, and nobody was ready for it.  During the two week scramble, our release pipeline was halted for a shipping product, which created unnecessary management pressure.

I say all this to point out that "hey, it was easy for me" doesn't necessarily imply that everyone else will have the same experience. 

Now, maybe the lesson is to simply lock our version number for every @types package that we reference.  But I'll point out that my node_modules folder has 1,134 top-level package folders in it, and **@types/lodash** is the first SemVer violation that I recall encountering in over a year.  Something ain't right about that.

> No one is forcing you to use DefinitelyTyped. [...] In fact, your organization could publish 
> your types for lodash once and depend strictly on those if you must control them and keep 
> them in sync.

So we should embrace & extend all our typings, isolating ourselves in a parallel universe of packages?  This is somewhat strange advice.  :-)

The simplest solution would be to honor SemVer.  The ideal solution would be to track compile-time versioning separate from runtime versioning.  The most practical solution would be for core libraries to NOT be early adopters of new compiler features.  I'm not sure what the right answer is, but I am sure that something ain't right with the current model heheh.

**Comment by lukas-zech-software** [comment]
I think this just points out a major problem with the versioning of typings, or better how they are used.
The version of a package in the `@types` scope currently expresses two different things simultaneously:
1. The version of the packages the typings are written against in SemVer (at least sort of), in this case lodash@^4.14.x
2. The version of the typings itself, which is not not SemVer as only the fix version is ever increased to maintain point 1

This structure is unable to express any real dependency for the typings as breaking changes cannot be documented, as we just saw. It can only express distinct versions without any information about compatibility towards a previous version of the typings.

If the typings would have their own SemVer version expressed in the package in could not express the lodash version it is written against anymore.

So either we introduce a second version to split these two responsibilities, although I have no idea how to consume this second version, or we accept that we cannot use SemVer for typings.

Currently I see only the second solution as practicable and would suggest, that typings should always be installed with a fixed version and must be updated manually.

**Comment by lukas-zech-software** [comment]
@pgonzal I definitely understand your concerns and frustration with this problem, but I think delaying the usage of new Typescript version cannot be the solution as there will always be projects using older version and determine the time to switch can only be arbitrarily chosen.

What is actually need is a sort of _long term support_ for typings which, at least in my opinion, can be easily achieved by fixing the version of the typings

**Comment by NPrice99** [comment]
If I could put in my two penny's worth on this one. I am using SPFX - I have no understanding of what goes on underneath SPFX.  But I do expect, rightly or wrongly, that when I compile my SPFX code that it does not break due to something totally beyond my control.  We all have deadlines to meet and if @pgonzal had not posted on my issue in SPFX I would never had worked out that this was the problem.  So I would have spent a load of wasted time trying to find the problem.  So I ask that if people make changes, that they are thoroughly tested and pass all of the gates / hurdles that they need to before being pulled into the master branch so that when the likes of me, and people like me, don't find problems which they have not caused and cannot fix.

**Comment by ericanderson** [comment]
It seems this complaint would be better lodged against the compiler which I believe makes claims of back-compat while DT does not.

**Comment by lukas-zech-software** [comment]
@NPrice99 As @ericanderson pointed out, this is the wrong place for your concerns.

The PullRequest met all requirements, was reviewed and passed all tests in this repository, therefore the changes introduced it this PR can hardly be blamed for the inconvenience it caused.

The problems arose from an misconception in the dependency management used in SPFX (and almost all other projects, including all of mine). Relying on typings using SemVer and expressing the dependencies to them with a weak version like  `~4.14.38` simply cannot work as DT does not and cannot follow SemVer as pointed out above.

It is unfortunate that this was uncovered now but it simply shows, that most of the community (including myself) relied on something that was never intended to work like that.

It was furthermore caused by the Compiler itself not following SemVer so that the change from 2.0 to 2.1 is not backward compatible

We will need to change the how we install typings and start adding them with fixed versions as these kind of changes will continue to break builds when new version of typings are installed automatically instead of reviewing the changes is this version manually

**Comment by pgonzal** [comment]
> The problems arose from a misconception in the dependency management used
> in SPFX (and almost all other projects, including all of mine). Relying on typings
> using SemVer and expressing the dependencies to them with a weak version
> like ~4.14.38 simply cannot work as DT does not and cannot follow SemVer as
> pointed out above.

Are there official docs somewhere communicating that **@types**-scoped NPM packages do not follow SemVer, by design?  That's news to me.

Even if what you say is true, what does it imply for libraries written natively in TypeScript?  Should they do a major SemVer increment when the upgrade their compiler?  For our internal projects, different people came to different conclusions.  Some did a major SemVer update, others coordinated with each other to migrate in a safe order.

If the open source community is going to rely heavily on TypeScript, we need some kind of agreement of what these version numbers mean so that we can avoid breaking end users.

Personally, I don't see why it's so important for the **@types** version to exactly match the underlying library. The purpose of the NPM package version is to control the NPM install algorithm, not to be a free-form field for conveying scenario specific information.   The **package.json** file format is [open-ended](http://stackoverflow.com/questions/10065564/add-custom-metadata-or-config-to-package-json-is-it-valid), so you could simply add a custom field such as "underlyingVersion" to indicate this information.

Lastly, I'd like to mention that @andy-ms's "opt-out" tag `@types/lodash@ts2.0` is useless to us, since we can't rewrite our entire node_modules tree to use that tag.  But an "opt-in" tag `@types/lodash@ts2.1` would probably be perfectly acceptable to the early adopters and small time projects that want to try out the new keywords.  I'm not seeking to hold back the wheels of progress indefinitely, but 2 months is a pretty short runway for people with lots of packages to shepherd, and who probably care more about stability than access to the latest compiler keywords.  (To that point, someone just came by my office this morning to say that the `Object.setPrototypeOf(this, FooError.prototype);` workaround still hasn't solved our issues with `Error` subclasses.)

@NPrice99 @ericanderson @mhegazy

**Comment by andrewconnell** [comment]
> @pgonzal - Personally, I don't see why it's so important for the @types version to exactly match the underlying library. The purpose of the NPM package version is to control the NPM install algorithm, not to be a free-form field for conveying scenario specific information. The package.json file format is open-ended, so you could simply add a custom field such as "underlyingVersion" to indicate this information.

Because a **@types** package has a specific version that's written for a specific version of a library it's defining. Ambient type definitions are just like documentation... they have to have strict versioning. If the underlying library changes, then the @types package can get updated to reflect the new version, but those who are on older versions can still rely on them.

FWIW, the system is working just fine... the problem is the author who took a downstream dependency seems to have made the incorrect assumption that DT conforms to symver which it doesn't. Yet another reason why it's a good practice when taking dependencies, you should specify explicit versions.

**Comment by pgonzal** [comment]
Suppose we have three different libraries, each in its own Git repo, like this:

```
my-button 1.0.0:
  @types/react: "0.14.46"

my-dialog 1.0.0:
  my-button: "^1.0.0"
  @types/react: "0.14.46"

my-app 1.0.0:
  my-dialog: "^1.0.0"
  @types/react: "0.14.46"
```

In this example, the `MyButton` class inherits from `React.Component`, and `MyDialog` exposes `MyButton` as properties.

Now, as you say, suppose that **@types/react** does not honor SemVer, so I cannot use "~" or "^" patterns.  And let's say that someone releases a version 0.14.47 which is a tiny, innocuous fix to a comment somewhere in the **@types/react** definitions:

```
my-button 1.0.0:
  @types/react: "0.14.47"  <-- tiny fix

my-dialog 1.0.0:  <-- broken  :-(
  my-button: "^1.0.0"
  @types/react: "0.14.46"

my-app 1.0.0:  <-- broken :-(
  my-dialog: "^1.0.0"
  @types/react: "0.14.46"
```

Why are my projects broken!?  It's because e.g. when compiling **my-dialog**, TypeScript will see two different versions of `React.Component`, which are incompatible from its perspective.

My only option is to start with a PR like this:

```
my-button 2.0.0:  <--- effect
  @types/react: "0.14.47"  <--- cause

my-dialog 1.0.0:
  my-button: "^1.0.0"
  @types/react: "0.14.46"

my-app 1.0.0:
  my-dialog: "^1.0.0"
  @types/react: "0.14.46"
```

Then I will do a PR to migrate **my-dialog**:

```
my-button 2.0.0:
  @types/react: "0.14.47"

my-dialog 2.0.0:  <-- effect
  my-button: "^2.0.0"  <-- upgraded
  @types/react: "0.14.46"

my-app 1.0.0:
  my-dialog: "^1.0.0"
  @types/react: "0.14.46"
```

...and finally I can do a third PR to migrate **my-app**.  This expensive, disruptive process is exactly what we must do when there is an actual breaking change.  That's life.

But the smart reader will notice that it's completely unnecessary for:

- **patch changes:**   fix bugs, but don't change any API contracts  (forwards/backwards compatible)
- **minor changes:**  add new stuff, but don't change existing stuff  (forwards compatible)

SemVer was invented exactly to avoid churn in these cases.  In this light, it seems really strange to designate **@types** as a special case, where every single modification must require the expensive overhead of a major breaking change.

Are we sure that's the approach we chose?  I haven't been able to find any official docs about @types versioning.

**Comment by ericanderson** [comment]
SemVer is not the defacto for the world. A project does not have to declare it is not semver. A project should declare that it is. 

I don't run your project but if I were facing this situation I would create a commons or shared package that all my packages depend on so that I wasn't in dependency hell. 

As for error subclassing, it's always been a bit broken in typescript. Have you considered writing what you need in JS and then declare your own types?

**Comment by andrewconnell** [comment]
@pgonzal There are no docs for @types versioning... because there is no such thing. @types is simply a NPM scope. They are all based on the versions of specific type definitions in DT (http://definitelytyped.org/)... which has no versioning stance ATM.

IMHO, you can't force another library to adopt & adhere to symver. I personally prefer it but you can't force DT to adopt it as it's been around for a very long time with tons of users working just fine.

**Comment by andrewconnell** [comment]
As it's been stated above, this isn't the right place for this discussion... spin up a new issue within DT if it's a DT versioning issue. This PR & the author of it is legit... its a root issue by someone who took a dependency from it.

**Comment by lukas-zech-software** [comment]
@pgonzal I don't quite understand your example
```
my-button 2.0.0:  <--- effect
  @types/react: "0.14.47"  <--- tiny fix
```

Why would you change the version of my-button if you update its types? If you really need to update the types, then you'll need to compile `my-button` with TS2.1 from now on, but this won't have any effect on any other package as consumers do not care how a package was compiled as long as the runtime code stays functional and can be consumed with the same API as before ...
So no reason to bump the major version of `my-button`

`my-dialog 1.0.0` could even be compiled with TS 1.8 while `my-button` uses TS2.1 and they'd still be compatible on Javascript level which is where they actually interact.


Anyway, for productive code it is *highly advisable* to lock down all dependencies using `npm shrinkwrap` which would not only circumvent this problem but also all issues with packages that do not follow SemVer, intentionally or by accident.
And NPM is full of such packages see [this issue at underscore](https://github.com/jashkenas/underscore/issues/1805) for a very prominent example of denying the usage of SemVer

**Comment by pgonzal** [comment]
@lukas-zech-software 

> If you really need to update the types, then you'll need to compile my-button with TS2.1
> from now on, but this won't have any effect on any other package as consumers

Sorry about the confusion. The example I gave does not involve a compiler upgrade.  You can assume all steps are built using the same compiler version.  

The point of this example was to illustrate the developer experience for a package that does not follow SemVer.  I was replying to Andrew's statement that DT is "working just fine" when it breaks SemVer.  In our experience, packages that break SemVer are painful to use.  I was suggesting that this pain might be a bigger concern than the goal of having the same NPM version number for **@types/react** and **react**.

> Why would you change the version of my-button if you update its types?

The second code block shows why.  If we don't bump **my-button** to 2.0.0, then `MyDialog` will fail to compile, because it is trying to render a `MyButton` which extends version 0.14.47 of `React.Component`, whereas `MyDialog` extends version 0.14.46 of `React.Component`.  They are effectively two different types, which happen to have the same name.  IIRC the error is something confusing like "Argument of type 'React.Component' is not assignable to parameter of type 'React.Component'"

> Anyway, for productive code it is highly advisable to lock down all dependencies using
> npm shrinkwrap which would not only circumvent this problem but also all issues with
> packages that do not follow SemVer, intentionally or by accident.

If your node_modules folder is already in a good state, then "npm shrinkwrap" is a useful way to reproduce that state later.  But in this scenario, we are trying to install a bunch of new package versions to get into a good state.  "npm shrinkwrap" doesn't help that problem.

Well, it might help if you're willing to manually edit the **npm-shrinkwrap.json** file.  This could work for simple cases, e.g. where we know that **@types/react** has no dependency changes from 0.14.46-> 0.14.47.  But in general, "npm install" is solving a complex multivariable SemVer equation that I would not feel comfortable advising people to tamper with.  My **npm-shrinkwrap.json** contains 390 kilobytes of JSON!

**Pull request merged**